### ISSUE TITLE
False positives with dataclasses

### ISSUE THREAD

**Issue by bretttully** [PyCQA/pylint#5857]
### Bug description

```python
"""repr=False causes false positives"""
import dataclasses


@dataclasses.dataclass(frozen=True)
class ExampleA:
    """Base class"""

    var_a: int


@dataclasses.dataclass(frozen=True, repr=False)
class ExampleB(ExampleA):
    """Child class with no repr"""

    var_b: int = 10


a = ExampleA(1)
assert a.var_a == 1

b = ExampleB(2)
assert b.var_a == 2
assert b.var_b == 10

b = ExampleB(2, 3)  # Too many positional arguments for constructor call
assert b.var_a == 2
assert b.var_b == 3

b = ExampleB(2, var_b=3)  # Unexpected keyword argument 'var_b' in constructor call
assert b.var_a == 2
assert b.var_b == 3
```

### Command used

```shell
pylint tmp.py
```

### Pylint output

```shell
************* Module tmp
tmp.py:26:4: E1121: Too many positional arguments for constructor call (too-many-function-args)
tmp.py:30:4: E1123: Unexpected keyword argument 'var_b' in constructor call (unexpected-keyword-arg)

------------------------------------------------------------------
Your code has been rated at 3.75/10 (previous run: 2.50/10, +1.25)
```

### Expected behavior

No warnings produced

### Pylint version

```shell
pylint 2.12.2
astroid 2.9.3
Python 3.8.12 | packaged by conda-forge | (default, Jan 30 2022, 23:42:07) 
[GCC 9.4.0]
```

### OS / Environment

Ubuntu 20.04, clean conda env

**Comment by Pierre-Sassoulas** [comment]
Closing as it's been fixed at some point before 2.15.0-dev0 or I can't reproduce.

**Comment by DanielNoord** [comment]
Fixed in https://github.com/PyCQA/astroid/pull/1764 😄

**Comment by alexey-pelykh** [comment]
A false-positive is happening when class is being decorated with a decorator like this:
```
def my_decorator(cls):
    # Check if called as @my_decorator()
    if cls is None:
        return my_decorator

    # Wrap as a dataclass
    cls = dataclass(
        cls,
        kw_only=True,
        frozen=True,  # NOTE: avoid dataclasses._field_assign using regular assignment
    )

    return cls
```

Though I'm not sure if such cases can even be handled.
